### DISCUSSION TITLE
12.3.1 tasmota-lite does not compile when using USE_SML_M

### DISCUSSION THREAD

**Discussion by gone-for-coding** [arendst/Tasmota#17544] [Q&A] [answered]
### PROBLEM DESCRIPTION
Compilation works fine when using `default_envs=tasmota` when `USE_SML_M` is defined but fails when using `default_envs=tasmota-lite`.

```
Linking .pio/build/tasmota-lite/firmware.elf
/workspace/Tasmota/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pio/build/tasmota-lite/src/tasmota.ino.cpp.o:(.text._Z8SML_Showb+0x28): undefined reference to `sml_hex_asci(unsigned int, char*)'
/workspace/Tasmota/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pio/build/tasmota-lite/src/tasmota.ino.cpp.o: in function `SML_Show(bool)':
tasmota.ino.cpp:(.text._Z8SML_Showb+0x17b): undefined reference to `sml_hex_asci(unsigned int, char*)'
collect2: error: ld returned 1 exit status
*** [.pio/build/tasmota-lite/firmware.elf] Error 1
```

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x ] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x ] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x ] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x ] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [ ] Device used (e.g., Sonoff Basic): _____
- [ x] Tasmota binary firmware version number used: _12.3.1____
  - [ ] Pre-compiled
  - [x ] Self-compiled
- [ ] Flashing tools used: _____
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:

### TO REPRODUCE
_Steps to reproduce the behavior:_

Set `user_config_override.h`
```
#ifndef USE_SCRIPT
	#define USE_SCRIPT
#endif
#ifndef USE_SML_M
	#define USE_SML_M
#endif
#ifdef USE_RULES
	#undef USE_RULES
#endif
```

Set `platformio_override.ini`
```
default_envs=tasmota-lite
```

Compile fails.

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

Compile succeeds like when not using tasmota-lite.

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by barbudor** [comment, marked as answer]
As expected and documented
You should not try to tweak predefined configurations
Only `tasmota`/`tasmota32` are meant to be customized

**Reply by gone-for-coding** [reply]
Thanks @barbudor! I read the manual over and over but forgot that piece during testing. Apparently I was somehow mislead by this one:

https://tasmota.github.io/docs/Smart-Meter-Interface/#special-commands
There is this tip which somehow made me think to adjust tasmota-lite ;)

> Also recommended, if you use lots of vars and increased buffer sizes to free up the image from unused drivers. You should get some inspiration from the tasmota-lite image definition in tasmota_configurations.h. 

And for reference for everyone looking at this post, here is the part of the documentation:
https://tasmota.github.io/docs/Compile-your-build/#customize-your-build

**Reply by sfromis** [reply]
It did not tell you to start from tasmota-lite, *only* to "get some inspiration" on how to trim features you have no use for. Anyway, trimming down is about selective use of `#undef`, but possibly a lot if you really need to squeeze it, and `my_user_config.h` comments does suggest how much space is taken by various features. As the main bottleneck on ESP8266, RAM is more important than flash space.

**Reply by gone-for-coding** [reply]
This is exactly how I did it now, with a rather lenghty list in `user_config_override.h` of `#ifdef` `#undef`.

**Reply by nikito7** [reply]
I keep "lite" def intact and overwrite after.
Maybe it's not the correct way, but it works.
Easy to update later.

https://github.com/nikito7/edpbox/blob/dev/tasmota/esp8266/build.h

**Reply by Jason2866** [reply]
@nikito7 This is not working, you may can activate some features. There are explicit #undefs in variant Tasmota-lite which are not possible to override. This is as wanted!

**Reply by nikito7** [reply]
I use "tasmota" build.

"tasmota lite custom" is something new. I not tested it yet.

**Comment by Jambola77** [comment]
Sorry, but I have the same problem and did not get the point. The documentation says to use:
```
#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif
#ifndef USE_SML_M
#define USE_SML_M
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif
```
But this leads to the error message above. What can I do to compile the binary tasmota-lite without errors?

**Reply by sfromis** [reply]
That you can do by not trying to add features to tasmota-lite.

As explained by barbudor, only the build environments `tasmota`/`tasmota32` are meant to be customized. That means that the other `tasmota-xxxx` environments do not support amendments of features.

**Reply by Jason2866** [reply]
Do as explained in the Docu. Link posted above.

**Reply by Jambola77** [reply]
Ah sorry, my error comes from tasmota-minimal not tasmota-lite. 
I have now wrapped the above defines as follows and this works:

```
#ifndef FIRMWARE_MINIMAL
  #ifndef USE_SCRIPT
  #define USE_SCRIPT
  #endif
  #ifndef USE_SML_M
  #define USE_SML_M
  #endif
  #ifdef USE_RULES
  #undef USE_RULES
  #endif
#endif
```

**Reply by sfromis** [reply]
Looks very different from what's explained in the documentation, as you should not touch anything near "minimal"

**Reply by Jambola77** [reply]
I added this to user_config_override.h. Nothing else was changed. 
Build envs are "tasmota" and "tasmota-minimal":
```
default_envs            =
; *** Uncomment the line(s) below to select version(s)
                          tasmota
;                          tasmota-debug
                          tasmota-minimal
[...]
```
Ok, I will only self-build tasmota in the future and using minimal from the repositories.

**Reply by sfromis** [reply]
It is fine to build minimal too, if you have a specific reason, but you do not get to add features to minimal. This could mean needing a separate build for minimal, if you need it (which you generally do not). I always use a precompiled minimal when upgrading my own builds.
